src/pulselistener/pulselistener/mercurial.py

```diff
@@ -0,0 +1,138 @@
+        _, self.ssh_key_path = tempfile.mkstemp(suffix='.key')
+        with open(self.ssh_key_path, 'w') as f:
+            f.write(ssh_key)
+
+        # Build ssh conf
+        conf = {
+            'StrictHostKeyChecking': 'no',
+            'User': ssh_user,
+            'IdentityFile': self.ssh_key_path,
+        }
+        self.ssh_conf = 'ssh {}'.format(' '.join('-o {}="{}"'.format(k, v) for k, v in conf.items())).encode('utf-8')
+
+        # Remove key when finished
+        atexit.register(self.cleanup)
+
+    def cleanup(self):
+        os.unlink(self.ssh_key_path)
+        logger.info('Removed ssh key')
+
+    async def run(self):
+        # Start by updating the repo
+        logger.info('Checking out tip', repo=self.repo_url)
+        self.repo = robust_checkout(self.repo_url, self.repo_dir)
+        self.repo.setcbout(lambda msg: logger.info('Mercurial', stdout=msg))
+        self.repo.setcberr(lambda msg: logger.info('Mercurial', stderr=msg))
+        logger.info('Initial clone finished')
+
+        # Wait for phabricator diffs to apply
+        while True:
+            diff = await self.queue.get()
+            assert isinstance(diff, dict)
+            assert 'phid' in diff
+
+            try:
+                await self.handle_diff(diff)
+
+            except hglib.error.CommandError as e:
+                logger.warn('Mercurial error on diff', error=e.err, args=e.args, phid=diff['phid'])
+
+                # Remove uncommited changes
+                self.repo.revert(self.repo_dir.encode('utf-8'), all=True)
+
+            except Exception as e:
+                logger.warn('Failed to process diff', error=e, phid=diff['phid'])
+
+                # Remove uncommited changes
+                self.repo.revert(self.repo_dir.encode('utf-8'), all=True)
+
+            # Notify the queue that the message has been processed
+            self.queue.task_done()
+
+    def clean(self):
+        '''
+        Steps to clean the mercurial repo
+        '''
+        logger.info('Remove all mercurial drafts')
+        try:
+            cmd = hglib.util.cmdbuilder(b'strip', rev=b'roots(outgoing())', force=True)
+            self.repo.rawcommand(cmd)
+        except hglib.error.CommandError as e:
+            if b'abort: empty revision set' not in e.err:
+                raise
+
+        logger.info('Pull updates from remote repo')
+        self.repo.pull()
+
+    async def handle_diff(self, diff):
+        '''
+        Handle a new diff received from Phabricator:
+        - apply revision to mercurial repo
+        - trigger push-to-try
+        '''
+        logger.info('Received diff {phid}'.format(**diff))
+
+        # Start by cleaning the repo
+        self.clean()
+
+        # Get the stack of patches
+        patches = self.phabricator_api.load_patches_stack(self.repo, diff, default_revision='central')
+        assert len(patches) > 0, 'No patches to apply'
+
+        # Get current revision
+        base = self.repo.tip()
```

Comment:
We should use the base of the stack (which is the changeset we are at right now, since load_patches_stack updates the repo to the base of the stack) rather than the tip of the repo (the tip is the latest commit, no?).

Reply:
Now `load_stack_patches()` gives the mercurial base revision it updated the repo to + the patches